README.md
# neocolor

Color definition, manipulation, conversion, and comparison in fluent Dart.

[![On pub.dev][pub_img]][pub_url]
[![Code coverage][cov_img]][cov_url]
[![Github action status][gha_img]][gha_url]
[![Dartdocs][doc_img]][doc_url]
[![Style guide][sty_img]][sty_url]

[pub_url]: https://pub.dartlang.org/packages/neocolor
[pub_img]: https://img.shields.io/pub/v/neocolor.svg
[gha_url]: https://github.com/neo-dart/neocolor/actions
[gha_img]: https://github.com/neo-dart/neocolor/workflows/Dart/badge.svg
[cov_url]: https://codecov.io/gh/neo-dart/neocolor
[cov_img]: https://codecov.io/gh/neo-dart/neocolor/branch/main/graph/badge.svg
[doc_url]: https://www.dartdocs.org/documentation/neocolor/latest
[doc_img]: https://img.shields.io/badge/Documentation-neocolor-blue.svg
[sty_url]: https://pub.dev/packages/neodart
[sty_img]: https://img.shields.io/badge/style-neodart-9cf.svg

## Purpose

Provides platform-agnostic (i.e. CLI, Flutter, Web) classes and conventions for
_defining_ (creating or parsing from popular formats), _manipulating_ (changing
elements like brightness, hue, saturation, etc), _conversion_ (from/to popular
formats), and _comparison_ (sometimes called "distance") `Color` elements.

## Usage

If you've used Flutter's [`Color`][dart-ui-color] class, you'll feel at home:

```dart
import 'package:neocolor/neocolor.dart';

void main() {
  const c1 = Color(0xFF42A5F5);
  const c2 = Color.fromARGB(0xFF, 0x42, 0xA5, 0xF5);
  print('$c1 == $c2: ${c1 == c2}');

  const c3 = Color.fromARGB(255, 66, 165, 245);
  const c4 = Color.fromRGBO(66, 165, 245, 1.0);
  print('$c3 == $c4: ${c3 == c4}');

  // NEW: Use HSB (sometimes called HSV) and HSL to create colors too!
  final c5 = Color.fromHSB(206.8, 0.7306, 0.9608);
  final c6 = Color.fromHSL(206.8, 0.8990, 0.6100);
  print('$c5 == $c6: ${c5 == c6}');
}
```

[dart-ui-color]: https://api.flutter.dev/flutter/dart-ui/Color-class.html

## Contributing

**This package welcomes [new issues][issues] and [pull requests][fork].**

[issues]: https://github.com/neo-dart/neocolor/issues/new
[fork]: https://github.com/neo-dart/neocolor/fork

Changes or requests that do not match the following criteria will be rejected:

1. Common decency as described by the [Contributor Covenant][code-of-conduct].
2. Making this library brittle/extensible by other libraries.
3. Adding platform-specific functionality.
4. A somewhat arbitrary bar of "complexity", everything should be _easy to use_.

[code-of-conduct]: https://www.contributor-covenant.org/version/1/4/code-of-conduct/

## Inspiration

Packages already exist that tackle this problem, often in similar ways:

- Flutter's [`dart:ui`][dart-ui-color]
- [`package:color`](https://pub.dev/packages/color)
- [`package:dynamic_color`](https://pub.dev/packages/dynamic_color)
- [`package:eo_color`](https://pub.dev/packages/eo_color)
- [`package:ansicolor`](https://pub.dev/packages/ansicolor)
- [`package:rainbow_color`](https://pub.dev/packages/rainbow_color)
- [`package:tinycolor2`](https://pub.dev/packages/tinycolor2)
- [`package:pigment`](https://pub.dev/packages/pigment)
- [`colord` (on NPM)](https://www.npmjs.com/package/colord)

If one of these packages suits your needs better, use it instead!


CHANGELOG.md
# CHANGELOG

## 0.1.0

**MISC**:

- Improve test coverage and tweaked the `README`.

**BREAKING CHANGES**:

- Removed value-type semantics from `HSLResult` and `HSBResult`. While useful,
  they defeated the main purpose of these classes - temporary objects to read
  individual `H/S/L/B` values.

## 0.0.0

- Initial release, just includes a single class: `Color`.


test/color_test.dart
// ignore_for_file: prefer_const_constructors

import 'package:neocolor/neocolor.dart';
import 'package:test/test.dart';

void main() {
  test('should extract alpha|red|green|blue from a 32-bit integer', () {
    final color = Color(0xAABBCCDD);

    expect(color.alpha, 0xAA);
    expect(color.red, 0xBB);
    expect(color.green, 0xCC);
    expect(color.blue, 0xDD);
  });

  test('should ignore bits > 32', () {
    final color = Color(0xFFAABBCCDD);

    expect(color, Color(0xAABBCCDD));
  });

  test('should create a color from RGB values', () {
    final color = Color.fromRGB(0xAA, 0xBB, 0xCC);

    expect(color, Color(0xFFAABBCC));
  });

  test('should align out of range RGB values', () {
    final color = Color.fromRGB(0xAAA, 0xBBB, 0xCCC);

    expect(color, Color(0xFFAABBCC));
  });

  test('should create a color from ARGB values', () {
    final color = Color.fromARGB(0xAA, 0xBB, 0xCC, 0xDD);

    expect(color, Color(0xAABBCCDD));
  });

  test('should align out of range ARGB values', () {
    final color = Color.fromARGB(0xAAA, 0xBBB, 0xCCC, 0xDDD);

    expect(color, Color(0xAABBCCDD));
  });

  test('should create a color from RGBO values', () {
    final color = Color.fromRGBO(0xBB, 0xCC, 0xDD, 1.0);

    expect(color, Color(0xFFBBCCDD));
  });

  test('should align out of range RGBO values', () {
    final color = Color.fromRGBO(0xBBB, 0xCCC, 0xDDD, 1.5);

    expect(color, Color(0x7EBBCCDD));
  });

  test('should create RGBA values from HSB', () {
    final color = Color.fromHSB(165, 0.50, 0.75, opacity: 0.5);

    expect(color, Color(0x7F60BFA7));
  });

  group('should create RGBA values from a range of HSB values', () {
    for (var h = 0.0; h < 360; h += 60) {
      test('<h=$h>', () {
        final color = Color.fromHSB(h, 0.50, 0.75, opacity: 0.5);
        expect(color.computeHSB().hue, h);
      });
    }
  });

  test('should create RGBA values from HSL', () {
    final color = Color.fromHSL(165, 0.50, 0.75, opacity: 0.5);

    expect(color, Color(0x7F9FDFCF));
  });

  group('should replace values withARGB', () {
    final start = Color(0xAABBCCDD);

    test('alpha', () {
      final color = start.withARGB(alpha: 0x99);

      expect(color, Color(0x99BBCCDD));
    });

    test('red', () {
      final color = start.withARGB(red: 0x99);

      expect(color, Color(0xAA99CCDD));
    });

    test('green', () {
      final color = start.withARGB(green: 0x99);

      expect(color, Color(0xAABB99DD));
    });

    test('blue', () {
      final color = start.withARGB(blue: 0x99);

      expect(color, Color(0xAABBCC99));
    });
  });

  group('should replace vaues withRGBO', () {
    final start = Color(0xAABBCCDD);

    test('opacity', () {
      final color = start.withRGBO(opacity: 1.0);

      expect(color, Color(0xFFBBCCDD));
    });

    test('red', () {
      final color = start.withRGBO(red: 0xFF);

      expect(color, Color(0xAAFFCCDD));
    });

    test('green', () {
      final color = start.withRGBO(green: 0xFF);

      expect(color, Color(0xAABBFFDD));
    });

    test('blue', () {
      final color = start.withRGBO(blue: 0xFF);

      expect(color, Color(0xAABBCCFF));
    });
  });

  test('should provide is{Opaque,Transparent,Visible} for alpha', () {
    final opaque = Color(0xFF000000);
    final transparent = opaque.withARGB(alpha: 0xCC);
    final hidden = opaque.withARGB(alpha: 0x00);

    expect(opaque.isOpaque, isTrue);
    expect(opaque.isTransparent, isFalse);
    expect(opaque.isVisible, isTrue);

    expect(transparent.isOpaque, isFalse);
    expect(transparent.isTransparent, isTrue);
    expect(transparent.isVisible, isTrue);

    expect(hidden.isVisible, isFalse);
  });

  test('should compute HSB/HSV', () {
    final hsb = Color(0x7F60BFA7).computeHSB();

    expect(hsb.hue, closeTo(164.8, 0.05));
    expect(hsb.saturation, closeTo(0.4974, 0.00005));
    expect(hsb.brightness, closeTo(0.7490, 0.00005));
    expect(hsb.opacity, 0x7F / 0xFF);
  });

  test('should compute HSB/HSV for grayscale RGB', () {
    final hsb = Color(0xFFCCCCCC).computeHSB();

    expect(hsb.hue, 0);
    expect(hsb.saturation, 0.0);
    expect(hsb.brightness, 0.8);
    expect(hsb.opacity, 1.0);
  });

  test('should compute HSL', () {
    final hsl = Color(0xFF42A5F5).computeHSL();

    expect(hsl.hue, closeTo(206.8, 0.05));
    expect(hsl.saturation, closeTo(0.899, 0.0005));
    expect(hsl.lightness, closeTo(0.6100, 0.0005));
    expect(hsl.opacity, 1.0);
  });

  test('should compute a hashCode', () {
    final a = Color.fromARGB(0x00, 0x33, 0x66, 0x99);
    final b = Color.fromARGB(0x00, 0x33, 0x66, 0x99);

    expect(a.hashCode, b.hashCode);
  });

  test('should return a readable toString()', () {
    final c = Color.fromARGB(0x00, 0x33, 0x66, 0x99);

    expect('$c', contains('0x00336699'));
  });
}


lib/neocolor.dart
import 'dart:math' as math;

import 'package:meta/meta.dart';

/// An immutable 32-bit color value in ARGB format.
///
/// This class is intended to be semantically similar to [`dart:ui`]'s `Color`,
/// but includes additional functionality, and is platform independent (i.e.
/// is suitable for non-Flutter environments, such as CLIs or non-Flutter web
/// applications).
///
/// [`dart:ui`]: https://api.flutter.dev/flutter/dart-ui/Color-class.html
///
/// **NOTE**: Unlike `dart:ui`, custom implementations are **not supported**.
/// Prefer [composition over inheritance][composition-over-inheritance].
///
/// [composition-over-inheritance]: https://en.wikipedia.org/wiki/Composition_over_inheritance
@immutable
@sealed
class Color {
  /// A 32-bit value representing this color.
  ///
  /// The bits are assigned as follows:
  /// - Bits 24-31 are the [alpha] value.
  /// - Bits 16-23 are the [red] value.
  /// - Bits 08-15 are the [green] value.
  /// - Bits 00-07 are the [blue] value.
  final int value;

  /// Constructs a color from the lower 32-bits of an [int].
  ///
  /// The bits are interpreted as follows:
  /// - Bits 24-31 are the [alpha] value.
  /// - Bits 16-23 are the [red] value.
  /// - Bits 08-15 are the [green] value.
  /// - Bits 00-07 are the [blue] value.
  ///
  /// In other words, if:
  /// - `AA` is the [alpha] value (in hex)
  /// - `RR` is the [red] value (in hex)
  /// - `GG` is the [green] value (in hex)
  /// - `BB` is the [blue] vlaue (in hex)
  ///
  /// ... the color can be expressed as `const Color(0xAARRGGBB`).
  ///
  /// For example, to get a fully opaque orange, you would use:
  /// ```
  /// //              red blue
  /// //              vv  vv
  /// const Color(0xFFFF9000)
  /// //            ^^  ^^
  ///               alpha^
  ///                   ^^
  ///                    green
  /// ```
  const Color(int value) : value = value & 0xFFFFFFFF;

  /// Constructs a _fully opaque_ color from the lower 8-bits of 3 integers.
  ///
  /// - [r] is [red], from `0` to `255`.
  /// - [g] is [green], from `0` to `255`.
  /// - [b] is [blue], from `0` to `255`.
  ///
  /// Out of range values are brought into range using modulo (`%`) `255.
  ///
  /// This constructor is semantically identical to [Color.fromARGB] where
  /// [alpha] is predefined as `255` (`0xFF`), suitable for fully opaque colors
  /// (i.e. where transparency is not intended or supported).
  const Color.fromRGB(
    int r,
    int g,
    int b,
  ) : value = ((0xFF << 24) |
                ((r & 0xFF) << 16) |
                ((g & 0xFF) << 8) |
                ((b & 0xFF) << 0)) &
            0xFFFFFFFF;

  /// Construct a color from the lower 8-bits of 4 integers.
  ///
  ///
  /// - [a] is [alpha], with `0` being transparent and `255` being fully opaque.
  /// - [r] is [red], from `0` to `255`.
  /// - [g] is [green], from `0` to `255`.
  /// - [b] is [blue], from `0` to `255`.
  ///
  /// Out of range values are brought into range using modulo (`%`) `255`.
  ///
  /// See also [Color.fromRGBO], which takes [alpha] as a floating point value.
  const Color.fromARGB(
    int a,
    int r,
    int g,
    int b,
  ) : value = (((a & 0xFF) << 24) |
                ((r & 0xFF) << 16) |
                ((g & 0xFF) << 8) |
                ((b & 0xFF) << 0)) &
            0xFFFFFFFF;

  /// Construct a color from the lower 8-bits of 3 integers and a [double].
  ///
  /// - [r] is [red], from `0` to `255`.
  /// - [g] is [green], from `0` to `255`.
  /// - [b] is [blue], from `0` to `255`.
  /// - [o] is [opacity], with `0.0` being transparent and `1.0` being fully
  ///   opaque.
  ///
  /// Out of range values are brought into range using modulo (`%`) `255`.
  ///
  /// See also [Color.fromARGB], which takes [opacity] as an integer value.
  const Color.fromRGBO(
    int r,
    int g,
    int b,
    double o,
  ) : value = ((((o * 0xff ~/ 1) & 0xff) << 24) |
                ((r & 0xff) << 16) |
                ((g & 0xff) << 8) |
                ((b & 0xff) << 0)) &
            0xFFFFFFFF;

  /// Constructs a color from _**H**ue-**S**aturation-**L**ightness_, an RGB
  /// alternative.
  ///
  /// HSL has a [cylindrical geometry][wiki-basic].
  ///
  /// - [h] is hue, or the perceived color in degrees (`0` to `360`).
  /// - [s] is saturation, or how "colorful" the color is (`0.0` to `1.0`).
  /// - [l] is lightness, or brightness of the color (`0.0` to `1.0`).
  ///
  /// Out of range values are brought into range using modulo (`%`).
  ///
  /// Unlike the other constructors (i.e. [Color.new] or [Color.fromRGB]) this
  /// constructor is computationally expensive and cannot be canonicalized
  /// (`const`). Where necessary, _consider_ storing the result, particularly
  /// for recurring values:
  ///
  /// ```
  /// while (true) {
  ///   // Performs a computation every loop.
  ///   drawColor(Color.fromHSL(0, 0.5, 0.5));
  /// }
  ///
  /// final color = Color.fromHSL(0, 0.5, 0.5);
  /// while (true) {
  ///   // Computation was already performed.
  ///   drawColor(color);
  /// }
  /// ```
  ///
  /// Optionally, provide [opacity] (defaults to fully opaque, or `1.0`).
  ///
  /// [wiki-basic]: https://en.wikipedia.org/wiki/HSL_and_HSV#Basic_principle
  factory Color.fromHSL(
    double h,
    double s,
    double l, {
    double opacity = 1.0,
  }) {
    // Converts to HSL->HSB which in turn is easier to convert to RGB.
    return HSLResult._(h, s, l, opacity).toColor();
  }

  /// Constructs a color from _**H**ue-**S**aturation-**B**rightness, an RGB
  /// alternative.
  ///
  /// HSV has a [cylindrical geometry][wiki-basic].
  ///
  /// - [h] is hue, or the perceived color in degrees (`0` to `360`).
  /// - [s] is saturation, or how "colorful" the color is (`0.0` to `1.0`).
  /// - [v] is brightness of the color (`0.0` to `1.0`).
  ///
  /// Out of range values are brought into range using modulo (`%`).
  ///
  /// Unlike the other constructors (i.e. [Color.new] or [Color.fromRGB]) this
  /// constructor is computationally expensive and cannot be canonicalized
  /// (`const`). Where necessary, _consider_ storing the result, particularly
  /// for recurring values:
  ///
  /// ```
  /// while (true) {
  ///   // Performs a computation every loop.
  ///   drawColor(Color.fromHSV(0, 0.5, 0.5));
  /// }
  ///
  /// final color = Color.fromHSV(0, 0.5, 0.5);
  /// while (true) {
  ///   // Computation was already performed.
  ///   drawColor(color);
  /// }
  /// ```
  ///
  /// Optionally, provide [opacity] (defaults to fully opaque, or `1.0`).
  ///
  /// [wiki-basic]: https://en.wikipedia.org/wiki/HSL_and_HSV#Basic_principle
  factory Color.fromHSB(
    double h,
    double s,
    double v, {
    double opacity = 1.0,
  }) {
    h = (h / 360) * 6;
    final hh = h.floor();
    final b = v * (1 - s);
    final c = v * (1 - (h - hh) * s);
    final d = v * (1 - (1 - h + hh) * s);
    final m = hh % 6;
    return Color._fromHSBToRGB(v, c, b, d, m, opacity);
  }

  factory Color._fromHSBToRGB(
    double v,
    double c,
    double b,
    double d,
    int m,
    double opacity,
  ) {
    final double red;
    final double green;
    final double blue;
    switch (m) {
      case 0:
        red = v;
        green = d;
        blue = b;
        break;
      case 1:
        red = c;
        green = v;
        blue = b;
        break;
      case 2:
        red = b;
        green = v;
        blue = d;
        break;
      case 3:
        red = b;
        green = c;
        blue = v;
        break;
      case 4:
        red = d;
        green = b;
        blue = v;
        break;
      case 5:
        red = v;
        green = b;
        blue = c;
        break;
      // coverage:ignore-start
      default:
        throw Error(/* I can't "prove" this, but this can never occur! */);
      // coverage:ignore-end
    }
    return Color.fromRGBO(
      _angleToColor(red),
      _angleToColor(green),
      _angleToColor(blue),
      opacity,
    );
  }

  static int _angleToColor(double c) => (c * 255).round();

  /// Returns a copy of color element, with non-null values replaced.
  ///
  /// - [alpha] is from `0` to `255`.
  /// - [red] is from `0` to `255`.
  /// - [green] is from `0` to `255`.
  /// - [blue] is from `0` to `255`.
  ///
  /// Out of range values are brought into range using modulo (`%`) `255`.
  ///
  /// This method is semantically identical to using [Color.fromARGB], i.e.:
  /// ```
  /// void example(Color a) {
  ///   final b = a.withARGB(red: 255);
  ///   final c = Color.fromARGB(a.alpha, 255, a.green, a.blue);
  ///
  ///   // true
  ///   print(b == c);
  /// }
  /// ```
  @useResult
  Color withARGB({int? alpha, int? red, int? green, int? blue}) {
    return Color.fromARGB(
      alpha ?? this.alpha,
      red ?? this.red,
      green ?? this.green,
      blue ?? this.blue,
    );
  }

  /// Returns a copy of color element, with non-null values replaced.
  ///
  /// - [opacity] is from `0.0` to `1.0`.
  /// - [red] is from `0` to `255`.
  /// - [green] is from `0` to `255`.
  /// - [blue] is from `0` to `255`.
  ///
  /// Out of range values are brought into range using modulo (`%`) `255`.
  ///
  /// This method is semantically identical to using [Color.fromRGBO], i.e.:
  /// ```
  /// void example(Color a) {
  ///   final b = a.withRGBO(red: 255);
  ///   final c = Color.fromRGBO(255, a.green, a.blue, a.opacity);
  ///
  ///   // true
  ///   print(b == c);
  /// }
  /// ```
  @useResult
  Color withRGBO({int? red, int? green, int? blue, double? opacity}) {
    return Color.fromRGBO(
      red ?? this.red,
      green ?? this.green,
      blue ?? this.blue,
      opacity ?? this.opacity,
    );
  }

  @override
  int get hashCode => value.hashCode;

  @override
  bool operator ==(Object other) {
    assert(
      other is! Color || other.runtimeType == Color,
      'Subtypes of Color are not permitted',
    );
    return other is Color && value == other.value;
  }

  /// The alpha channel of this color as an 8-bit value.
  ///
  /// A value of `0` means this color is fully transparent.
  /// A value of `255` means this color is fully opaque.
  ///
  /// See also: [isTransparent], [isVisible] and [isOpaque].
  int get alpha => (0xFF000000 & value) >> 24;

  /// The alpha channel of this color as a [double].
  ///
  /// A value of `0.0` means this color is fully transparent.
  /// A value of `1.0` means this color is fully opaque.
  ///
  /// See also: [isTransparent], [isVisible], and [isOpaque].
  double get opacity => alpha / 0xFF;

  /// Whether the color is _fully_ opaque, i.e. `alpha == 0xFF`.
  bool get isOpaque => alpha == 0xFF;

  /// Whether the color is **not** _fully_ opaque, i.e. `alpha < 0xFF`.
  bool get isTransparent => alpha < 0xFF;

  /// Whether the color is **not** _fully_ transparent, i.e. `alpha > 0x00`.
  bool get isVisible => alpha > 0x00;

  /// The red channel of this color as an 8-bit value.
  int get red => (0x00FF0000 & value) >> 16;

  /// The green channel of this color as an 8-bit value.
  int get green => (0x0000FF00 & value) >> 8;

  /// The blue channel of this color as an 8-bit value.
  int get blue => (0x000000FF & value) >> 0;

  /// Computes and returns as hue, saturation, and brightness.
  ///
  /// This method is computationally expensive, and should be used sparingly.
  ///
  /// See also: [computeHSL].
  @useResult
  HSBResult computeHSB() {
    final r = red / 255;
    final g = green / 255;
    final b = blue / 255;

    final min = math.min(math.min(r, g), b);
    final max = math.max(math.max(r, g), b);

    // Black-Gray-White
    if (min == max) {
      return HSBResult._(0, 0, min, opacity);
    }

    // Colors other than Black-Gray-White.
    final double d;
    final double h;

    if (min == r) {
      d = g - b;
      h = 3;
    } else if (min == b) {
      d = r - g;
      h = 1;
    } else {
      d = b - r;
      h = 5;
    }

    return HSBResult._(
      60.0 * (h - d / (max - min)),
      (max - min) / max,
      max,
      opacity,
    );
  }

  /// Computes and returns as hue, saturation, and lightness.
  ///
  /// This method is computationally expensive, and should be used sparingly.
  ///
  /// See also: [computeHSB].
  @useResult
  HSLResult computeHSL() => computeHSB().toHSL();

  @override
  String toString() {
    return 'Color <0x${value.toRadixString(16).toUpperCase().padLeft(8, '0')}>';
  }
}

/// Computational result of converting an RGBA [Color] to HSB.
@immutable
@sealed
class HSBResult {
  /// Perceived color in degrees (`0` to `360`).
  final double hue;

  /// How "colorful" the color is (`0.0` to `1.0`).
  final double saturation;

  /// Brightness of the color (`0.0` to `1.0`).
  final double brightness;

  /// The alpha channel of this color as a [double].
  ///
  /// A value of `0.0` means this color is fully transparent.
  /// A value of `1.0` means this color is fully opaque.
  final double opacity;

  /// All instances should be created within `package:neocolor` _only_.
  ///
  /// While it would be possible to make this class' constructor public, keeping
  /// it private and denoting the contract of this class as _sealed_ means we
  /// avoid having to bounds-check any of the values. In other words, values
  /// are guaranteed to be valid in a well-behaving application.
  ///
  /// It's also trivial to create instances of this class (i.e. for testing):
  /// ```
  /// // A bit wordy, but does the job alright.
  /// Color.fromHSB(h, s, b).toHSB();
  /// ```
  const HSBResult._(
    this.hue,
    this.saturation,
    this.brightness,
    this.opacity,
  );

  /// Converts back to an RGBA [Color].
  ///
  /// This is semantically identical to [Color.fromHSB], i.e.:
  /// ```
  /// void example(HSBResult hsb) {
  ///   final a = Color.fromHSB(hsl.hue, hsl.saturation, hsl.brightness);
  ///   final b = result.toColor();
  ///
  ///   // true
  ///   print(a == b);
  /// }
  /// ```
  @useResult
  Color toColor() {
    return Color.fromHSB(
      hue,
      saturation,
      brightness,
      opacity: opacity,
    );
  }

  /// Converts to _**H**ue-**S**aturation-**L**ightness_.
  ///
  /// See [HSLResult] and [Color.fromHSL] for details.
  @useResult
  HSLResult toHSL() {
    final h = hue;
    final s = saturation;
    final v = brightness;
    final l = (2.0 - s) * v;
    return HSLResult._(
      h,
      l == 0 || l == 2.0 ? 0 : s * v / (l <= 1.0 ? l : 2 - l),
      l * 5 / 10,
      opacity,
    );
  }
}

/// Computational result of converting an RGBA [Color] to HSL.
@immutable
@sealed
class HSLResult {
  /// Perceived color in degrees (`0` to `360`).
  final double hue;

  /// How "colorful" the color is (`0.0` to `1.0`).
  final double saturation;

  /// Brightness of the color (`0.0` to `1.0`).
  final double lightness;

  /// The alpha channel of this color as a [double].
  ///
  /// A value of `0.0` means this color is fully transparent.
  /// A value of `1.0` means this color is fully opaque.
  final double opacity;

  /// All instances should be created within `package:neocolor` _only_.
  ///
  /// While it would be possible to make this class' constructor public, keeping
  /// it private and denoting the contract of this class as _sealed_ means we
  /// avoid having to bounds-check any of the values. In other words, values
  /// are guaranteed to be valid in a well-behaving application.
  ///
  /// It's also trivial to create instances of this class (i.e. for testing):
  /// ```
  /// // A bit wordy, but does the job alright.
  /// Color.fromHSL(h, s, l).toHSL();
  /// ```
  const HSLResult._(
    this.hue,
    this.saturation,
    this.lightness,
    this.opacity,
  );

  /// Converts back to an RGBA [Color].
  ///
  /// This is semantically identical to [Color.fromHSL], i.e.:
  /// ```
  /// void example(HSLResult hsl) {
  ///   final a = Color.fromHSL(hsl.hue, hsl.saturation, hsl.lightness);
  ///   final b = result.toColor();
  ///
  ///   // true
  ///   print(a == b);
  /// }
  /// ```
  @useResult
  Color toColor() => toHSB().toColor();

  /// Converts to _**H**ue-**S**aturation-**B**rightness_.
  ///
  /// See [HSBResult] and [Color.fromHSB] for details.
  @useResult
  HSBResult toHSB() {
    final h = hue;
    final l = lightness;
    final s = saturation * (l < 0.5 ? l : 1 - l);
    final o = opacity;
    return HSBResult._(
      h,
      s > 0 ? ((2 * s) / (l + s)) : 0,
      l + s,
      o,
    );
  }
}
